packages/compass-components/src/hooks/use-toast.tsx

```diff
@@ -0,0 +1,168 @@
+  body: React.ReactNode;
+  variant: ToastVariant;
+  progress?: number;
+  timeout?: number;
+};
+
+type ToastState = ToastProperties & {
+  timeoutRef?: ReturnType<typeof setTimeout>;
+};
+
+type ToastId = string;
+
+type ToastContextState = {
+  toasts: Record<ToastId, ToastState>;
+  setToasts: (toasts: Record<ToastId, ToastState>) => void;
+};
+
+interface ToastActions {
+  openToast: (id: ToastId, toastProperties: ToastProperties) => void;
+  closeToast: (id: ToastId) => void;
+}
+
+const ToastContext = createContext<ToastContextState>({
+  toasts: {},
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  setToasts: (t: Record<ToastId, ToastProperties>) => {
+    return;
+  },
+});
+
+const toastStyles = css({
+  button: {
+    position: 'absolute',
+  },
+});
+
+const ToastPile = (): React.ReactElement => {
+  const { toasts } = useContext(ToastContext);
+  const { closeToast } = useGlobalToast();
+
+  return (
+    <>
+      {Object.entries(toasts).map(
+        ([id, { title, body, variant, progress }]) => (
+          <Toast
+            className={toastStyles}
+            key={id}
+            title={title}
+            body={body}
+            variant={variant}
+            progress={progress}
+            open={true}
+            close={() => closeToast(id)}
+          />
+        )
+      )}
+    </>
+  );
+};
+
+/**
+ * @example
+ *
+ * ```
+ * const MyButton = () => {
+ *   const { openToast } = useToast('my-namespace');
+ *   return <button onClick={() => openToast(
+ *      'myToast1', {title: 'This is a notification'})} />
+ * };
+ *
+ * <ToastArea><MyButton/><ToastArea>
+ * ```
+ *
+ * @returns
+ */
+export const ToastArea = ({
+  children,
+}: {
+  children?: React.ReactNode;
+}): React.ReactElement => {
+  const [toasts, setToasts] = useState({});
+
+  return (
+    <ToastContext.Provider value={{ toasts, setToasts }}>
+      <>{children}</>
+      <ToastPile />
+    </ToastContext.Provider>
+  );
+};
+
+function useGlobalToast(): ToastActions {
+  const { toasts, setToasts } = useContext(ToastContext);
+
+  const closeToast = useCallback(
+    (toastId: ToastId): void => {
+      const { timeoutRef } = toasts[toastId] || {};
+      if (timeoutRef) {
+        clearTimeout(timeoutRef);
+      }
+
+      const newToasts = { ...toasts };
+      delete newToasts[toastId];
+      setToasts(newToasts);
+    },
+    [toasts, setToasts]
+  );
+
+  const openToast = useCallback(
+    (toastId: ToastId, toastProperties: ToastProperties): void => {
+      // if updating clear timeouts first
+      const { timeoutRef } = toasts[toastId] || {};
+      if (timeoutRef) {
+        clearTimeout(timeoutRef);
+      }
+
+      setToasts({
+        ...toasts,
+        [toastId]: {
+          ...toastProperties,
+          timeoutRef: toastProperties.timeout
+            ? setTimeout(() => {
+                closeToast(toastId);
+              }, toastProperties.timeout)
+            : undefined,
+        },
+      });
```

Comment:
Because `setState` is not a sync operation, running `setToasts` like that will lead to concurrently opened toasts swallowing each other (you can test it by calling `openToast` in sequence anywhere in the code), what you want here is:

```suggestion
      setToasts(currentState => ({
        ...currentState,
        [toastId]: {
          ...toastProperties,
          timeoutRef: toastProperties.timeout
            ? setTimeout(() => {
                closeToast(toastId);
              }, toastProperties.timeout)
            : undefined,
        }
      }));
```

or you can move this logic into a React reducer with explicit actions for put / delete action types so that you ensure that you are always deriving state from the previous one, not overriding it.

We were just chatting with Basit about this, my rule of thumb is that if you need to put an complex data structure to `useState`, you usually either want to just split it to multiple primitive values that are controlled by `useState`, or use `useReducer` instead

Reply:
That's my noobness revealing itself :), nice to know.